File: translated/f-to-c/DRB067-restrictpointer1-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Array initialization using assignments. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

void foo(double** newSxx, double** newSyy, int len)
{
    int i;
    double* tar1;
    double* tar2;

    *newSxx = (double*)malloc(len * sizeof(double));
    *newSyy = (double*)malloc(len * sizeof(double));

    tar1 = (double*)malloc(len * sizeof(double));
    tar2 = (double*)malloc(len * sizeof(double));

    *newSxx = tar1;
    *newSyy = tar2;

    #pragma omp parallel for private(i) firstprivate(len)
    for (i = 0; i < len; i++) {
        tar1[i] = 0.0;
        tar2[i] = 0.0;
    }

    printf("%f %f\n", tar1[len-1], tar2[len-1]);
    
    // In C, newSxx and newSyy are pointing to tar1 and tar2,
    // so we only need to free tar1 and tar2
    free(tar1);
    free(tar2);
}

int main()
{
    int len = 1000;
    double* newSxx;
    double* newSyy;

    newSxx = (double*)malloc(len * sizeof(double));
    newSyy = (double*)malloc(len * sizeof(double));

    foo(&newSxx, &newSyy, len);

    // newSxx and newSyy now point to memory allocated inside foo(),
    // which has already been freed
    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB067-restrictpointer1-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Array initialization using assignments. No data race pairs.

package main

import (
	"fmt"
	"runtime"
	"sync"
)

// Module DRB067 translated to package-level functions
func foo(newSxx, newSyy []float64, length int) {
	var tar1, tar2 []float64

	newSxx = make([]float64, length)
	newSyy = make([]float64, length)
	tar1 = make([]float64, length)
	tar2 = make([]float64, length)

	newSxx = tar1[:]
	newSyy = tar2[:]

	//$omp parallel do private (i) firstprivate (len)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg.Add(1)
		go func(start, end, length int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				tar1[i-1] = 0.0
				tar2[i-1] = 0.0
			}
		}(start, end, length)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("%f %f\n", tar1[length-1], tar2[length-1])
}

func main() {
	var length int = 1000
	var newSxx, newSyy []float64

	newSxx = make([]float64, length)
	newSyy = make([]float64, length)

	foo(newSxx, newSyy, length)
}
